File: ios/Packages/GemstoneServices/Sources/Stores/SupportStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import protocol Gemstone.GemSupportStore
import typealias Gemstone.SupportMessage
import typealias Gemstone.SupportTyping
import GemstonePrimitives
import Primitives
import Store

@Observable
public final class GemstoneSupportStore: GemSupportStore, @unchecked Sendable {
    @ObservationIgnored
    private let store: SupportChatStore
    @ObservationIgnored
    private let lock = NSLock()
    @ObservationIgnored
    private var storedAgent: SupportAgent?

    public init(store: SupportChatStore) {
        self.store = store
    }

    public var typingAgent: SupportAgent? {
        access(keyPath: \.typingAgent)
        return lock.withLock { storedAgent }
    }

    public func saveMessages(messages: [Gemstone.SupportMessage]) async throws {
        try store.addMessages(messages.map { try Primitives.SupportMessage($0) })
    }

    public func saveMessage(id: String, message: Gemstone.SupportMessage) async throws {
        try store.replace(id: id, with: Primitives.SupportMessage(message))
    }

    public func updateTyping(typing: Gemstone.SupportTyping) throws {
        let typing = try Primitives.SupportTyping(typing)
        switch typing.status {
        case .on: setTypingAgent(typing.agent)
        case .off: setTypingAgent(.none)
        }
    }

    public func clearTyping() throws {
        setTypingAgent(.none)
    }

    public func clearTypingAgent() {
        setTypingAgent(.none)
    }

    private func setTypingAgent(_ agent: SupportAgent?) {
        withMutation(keyPath: \.typingAgent) {
            lock.withLock { storedAgent = agent }
        }
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stores/SupportStore.kt
package com.gemwallet.android.data.services.gemstone.stores

import com.gemwallet.android.data.services.store.database.SupportMessagesDao
import com.gemwallet.android.data.services.store.database.entities.toRecord
import com.gemwallet.android.ext.toPrimitives
import com.wallet.core.primitives.SupportAgent
import com.wallet.core.primitives.SupportMessageStatus
import com.wallet.core.primitives.SupportTypingStatus
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import uniffi.gemstone.GemSupportStore
import uniffi.gemstone.SupportMessage as GemSupportMessage
import uniffi.gemstone.SupportTyping as GemSupportTyping

class GemstoneSupportStore(private val supportMessagesDao: SupportMessagesDao) : GemSupportStore {

    private val agent = MutableStateFlow<SupportAgent?>(null)
    val typingAgent: StateFlow<SupportAgent?> = agent.asStateFlow()

    override fun updateTyping(typing: GemSupportTyping) {
        agent.value = typing.toPrimitives().let {
            when (it.status) {
                SupportTypingStatus.On -> it.agent
                SupportTypingStatus.Off -> null
            }
        }
    }

    override fun clearTyping() {
        agent.value = null
    }

    override suspend fun saveMessages(messages: List<GemSupportMessage>) {
        supportMessagesDao.addMessages(messages.map { it.toPrimitives().toRecord() })
    }

    override suspend fun saveMessage(id: String, message: GemSupportMessage) {
        supportMessagesDao.replace(id, message.toPrimitives().toRecord())
    }

    override suspend fun failPendingMessages(exceptIds: List<String>) {
        supportMessagesDao.failPending(
            sending = SupportMessageStatus.Sending.string,
            failed = SupportMessageStatus.Failed.string,
            exceptIds = exceptIds,
        )
    }
}
